File: common/src/main/java/org/apache/celeborn/common/write/PushState.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.celeborn.common.write;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;
import java.util.concurrent.atomic.LongAdder;

import org.apache.commons.lang3.tuple.Pair;

import org.apache.celeborn.common.CelebornConf;
import org.apache.celeborn.common.CommitMetadata;
import org.apache.celeborn.common.protocol.PartitionLocation;
import org.apache.celeborn.common.util.JavaUtils;

public class PushState {

  private final int pushBufferMaxSize;
  public AtomicReference<IOException> exception = new AtomicReference<>();
  private final InFlightRequestTracker inFlightRequestTracker;
  // partition id -> CommitMetadata
  private final ConcurrentHashMap<Integer, CommitMetadata> commitMetadataMap =
      JavaUtils.newConcurrentHashMap();

  private final Map<String, LocationPushFailedBatches> failedBatchMap;

  // Per map task write path stats, used to diagnose slow shuffle write.
  // Time a mapper thread is blocked by the push queue backpressure (DataPusher.addTask).
  private final LongAdder queueWaitTimeNanos = new LongAdder();
  // Time push threads are blocked by the in-flight limit (limitMaxInFlight).
  private final LongAdder inflightWaitTimeNanos = new LongAdder();
  // Time mapperEnd waits for all in-flight batches to be done (limitZeroInFlight).
  private final LongAdder drainWaitTimeNanos = new LongAdder();
  // Number of push batches whose round trip time exceeds the slow push threshold.
  private final LongAdder slowPushCount = new LongAdder();
  private final AtomicLong maxPushRttNanos = new AtomicLong(0);

  public PushState(CelebornConf conf) {
    pushBufferMaxSize = conf.clientPushBufferMaxSize();
    inFlightRequestTracker = new InFlightRequestTracker(conf, this);
    failedBatchMap = JavaUtils.newConcurrentHashMap();
  }

  public void addQueueWaitTime(long nanos) {
    queueWaitTimeNanos.add(nanos);
  }

  public void addInflightWaitTime(long nanos) {
    inflightWaitTimeNanos.add(nanos);
  }

  public void addDrainWaitTime(long nanos) {
    drainWaitTimeNanos.add(nanos);
  }

  public void recordPushRtt(long rttNanos, long slowThresholdNanos) {
    maxPushRttNanos.accumulateAndGet(rttNanos, Math::max);
    if (rttNanos > slowThresholdNanos) {
      slowPushCount.increment();
    }
  }

  public long getQueueWaitTimeMs() {
    return TimeUnit.NANOSECONDS.toMillis(queueWaitTimeNanos.sum());
  }

  public long getInflightWaitTimeMs() {
    return TimeUnit.NANOSECONDS.toMillis(inflightWaitTimeNanos.sum());
  }

  public long getDrainWaitTimeMs() {
    return TimeUnit.NANOSECONDS.toMillis(drainWaitTimeNanos.sum());
  }

  public long getSlowPushCount() {
    return slowPushCount.sum();
  }

  public long getMaxPushRttMs() {
    return TimeUnit.NANOSECONDS.toMillis(maxPushRttNanos.get());
  }

  public void cleanup() {
    inFlightRequestTracker.cleanup();
  }

  // key: ${primary addr}, ${replica addr} value: list of data batch
  public final ConcurrentHashMap<Pair<String, String>, DataBatches> batchesMap =
      JavaUtils.newConcurrentHashMap();

  public boolean addBatchData(
      Pair<String, String> addressPair, PartitionLocation loc, int batchId, byte[] body) {
    DataBatches batches = batchesMap.computeIfAbsent(addressPair, (s) -> new DataBatches());
    batches.addDataBatch(loc, batchId, body);
    return batches.getTotalSize() > pushBufferMaxSize;
  }

  public DataBatches takeDataBatches(Pair<String, String> addressPair) {
    return batchesMap.remove(addressPair);
  }

  public int nextBatchId() {
    return inFlightRequestTracker.nextBatchId();
  }

  public void addBatch(int batchId, int batchBytesSize, String hostAndPushPort) {
    inFlightRequestTracker.addBatch(batchId, batchBytesSize, hostAndPushPort);
  }

  public void removeBatch(int batchId, String hostAndPushPort) {
    inFlightRequestTracker.removeBatch(batchId, hostAndPushPort);
  }

  public void onSuccess(String hostAndPushPort) {
    inFlightRequestTracker.onSuccess(hostAndPushPort);
  }

  public void onCongestControl(String hostAndPushPort) {
    inFlightRequestTracker.onCongestControl(hostAndPushPort);
  }

  public boolean limitMaxInFlight(String hostAndPushPort) throws IOException {
    return inFlightRequestTracker.limitMaxInFlight(hostAndPushPort);
  }

  public boolean limitZeroInFlight() throws IOException {
    return inFlightRequestTracker.limitZeroInFlight();
  }

  public int remainingAllowPushes(String hostAndPushPort) {
    return inFlightRequestTracker.remainingAllowPushes(hostAndPushPort);
  }

  public void recordFailedBatch(String partitionId, int mapId, int attemptId, int batchId) {
    this.failedBatchMap
        .computeIfAbsent(partitionId, (s) -> new LocationPushFailedBatches())
        .addFailedBatch(mapId, attemptId, batchId);
  }

  public Map<String, LocationPushFailedBatches> getFailedBatches() {
    return this.failedBatchMap;
  }

  public int[] getCRC32PerPartition(boolean shuffleIntegrityCheckEnabled, int numPartitions) {
    if (!shuffleIntegrityCheckEnabled) {
      return new int[0];
    }

    int[] crc32PerPartition = new int[numPartitions];
    for (Map.Entry<Integer, CommitMetadata> entry : commitMetadataMap.entrySet()) {
      crc32PerPartition[entry.getKey()] = entry.getValue().getChecksum();
    }
    return crc32PerPartition;
  }

  public long[] getBytesWrittenPerPartition(
      boolean shuffleIntegrityCheckEnabled, int numPartitions) {
    if (!shuffleIntegrityCheckEnabled) {
      return new long[0];
    }
    long[] bytesWrittenPerPartition = new long[numPartitions];
    for (Map.Entry<Integer, CommitMetadata> entry : commitMetadataMap.entrySet()) {
      bytesWrittenPerPartition[entry.getKey()] = entry.getValue().getBytes();
    }
    return bytesWrittenPerPartition;
  }

  public void addDataWithOffsetAndLength(int partitionId, byte[] data, int offset, int length) {
    CommitMetadata commitMetadata =
        commitMetadataMap.computeIfAbsent(partitionId, id -> new CommitMetadata());
    commitMetadata.addDataWithOffsetAndLength(data, offset, length);
  }

  public void addData(int partitionId, ByteBuffer data) {
    CommitMetadata commitMetadata =
        commitMetadataMap.computeIfAbsent(partitionId, id -> new CommitMetadata());
    commitMetadata.addData(data);
  }
}

File: cpp/celeborn/client/writer/PushState.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "celeborn/client/writer/PushState.h"

namespace celeborn {
namespace client {

PushState::PushState(const conf::CelebornConf& conf)
    : waitInflightTimeoutMs_(conf.clientPushLimitInFlightTimeoutMs()),
      deltaMs_(conf.clientPushLimitInFlightSleepDeltaMs()),
      pushStrategy_(PushStrategy::create(conf)),
      maxInFlightReqsTotal_(conf.clientPushMaxReqsInFlightTotal()),
      maxInFlightBytesSizeEnabled_(
          conf.clientPushMaxBytesSizeInFlightEnabled()),
      maxInFlightBytesSizeTotal_(conf.clientPushMaxBytesSizeInFlightTotal()),
      maxInFlightBytesSizePerWorker_(
          conf.clientPushMaxBytesSizeInFlightPerWorker()) {
  if (maxInFlightBytesSizeEnabled_) {
    inflightBytesSizePerAddress_.emplace();
    inflightBatchBytesSizes_.emplace();
  }
}

int PushState::nextBatchId() {
  return currBatchId_.fetch_add(1);
}

void PushState::addBatch(
    int batchId,
    int batchBytesSize,
    const std::string& hostAndPushPort) {
  auto batchIdSet = inflightBatchesPerAddress_.computeIfAbsent(
      hostAndPushPort,
      [&]() { return std::make_shared<utils::ConcurrentHashSet<int>>(); });
  batchIdSet->insert(batchId);
  totalInflightReqs_.fetch_add(1);

  if (maxInFlightBytesSizeEnabled_) {
    auto bytesSizePerAddress = inflightBytesSizePerAddress_->computeIfAbsent(
        hostAndPushPort,
        [&]() { return std::make_shared<std::atomic<long>>(0); });
    bytesSizePerAddress->fetch_add(batchBytesSize);
    inflightBatchBytesSizes_->set(batchId, batchBytesSize);
    totalInflightBytes_.fetch_add(batchBytesSize);
  }
}

void PushState::onSuccess(const std::string& hostAndPushPort) {
  pushStrategy_->onSuccess(hostAndPushPort);
}

void PushState::onCongestControl(const std::string& hostAndPushPort) {
  pushStrategy_->onCongestControl(hostAndPushPort);
}

void PushState::removeBatch(int batchId, const std::string& hostAndPushPort) {
  auto batchIdSetOptional = inflightBatchesPerAddress_.get(hostAndPushPort);
  if (batchIdSetOptional.has_value()) {
    auto batchIdSet = batchIdSetOptional.value();
    batchIdSet->erase(batchId);
  } else {
    LOG(WARNING) << "BatchIdSet of " << hostAndPushPort << " doesn't exist.";
  }

  totalInflightReqs_.fetch_sub(1);

  if (maxInFlightBytesSizeEnabled_) {
    auto inflightBatchBytesSize = inflightBatchBytesSizes_->get(batchId);
    inflightBatchBytesSizes_->erase(batchId);
    if (inflightBatchBytesSize.has_value()) {
      auto inflightBytesSize =
          inflightBytesSizePerAddress_->get(hostAndPushPort);
      if (inflightBytesSize.has_value()) {
        inflightBytesSize.value()->fetch_sub(inflightBatchBytesSize.value());
      }
      totalInflightBytes_.fetch_sub(inflightBatchBytesSize.value());
    }
  }
}

bool PushState::limitMaxInFlight(const std::string& hostAndPushPort) {
  throwIfExceptionExists();

  pushStrategy_->limitPushSpeed(*this, hostAndPushPort);
  int currentMaxReqsInFlight =
      pushStrategy_->getCurrentMaxReqsInFlight(hostAndPushPort);

  auto batchIdSet = inflightBatchesPerAddress_.computeIfAbsent(
      hostAndPushPort,
      [&]() { return std::make_shared<utils::ConcurrentHashSet<int>>(); });
  std::shared_ptr<std::atomic<long>> batchBytesSize = nullptr;
  if (maxInFlightBytesSizeEnabled_) {
    batchBytesSize = inflightBytesSizePerAddress_->computeIfAbsent(
        hostAndPushPort,
        [&]() { return std::make_shared<std::atomic<long>>(0); });
  }
  long times = waitInflightTimeoutMs_ / deltaMs_;
  for (; times > 0; times--) {
    if (cleaned_) {
      return false;
    }

    bool reqCountWithinLimits =
        (totalInflightReqs_ <= maxInFlightReqsTotal_ &&
         static_cast<int>(batchIdSet->size()) <= currentMaxReqsInFlight);
    bool byteSizeWithinLimits = false;

    if (maxInFlightBytesSizeEnabled_ && batchBytesSize) {
      byteSizeWithinLimits =
          (totalInflightBytes_.load() <= maxInFlightBytesSizeTotal_ &&
           batchBytesSize->load() <= maxInFlightBytesSizePerWorker_);
    }

    if (reqCountWithinLimits ||
        (maxInFlightBytesSizeEnabled_ && byteSizeWithinLimits)) {
      break;
    }

    throwIfExceptionExists();
    std::this_thread::sleep_for(utils::MS(deltaMs_));
  }

  if (times <= 0) {
    if (totalInflightReqs_ > maxInFlightReqsTotal_ ||
        static_cast<int>(batchIdSet->size()) > currentMaxReqsInFlight) {
      LOG(WARNING) << "After waiting for " << waitInflightTimeoutMs_
                   << " ms, there are still " << totalInflightReqs_
                   << " requests in flight (limit: " << maxInFlightReqsTotal_
                   << "): " << batchIdSet->size()
                   << " batches in flight for hostAndPushPort "
                   << hostAndPushPort << ", which exceeds the current limit "
                   << currentMaxReqsInFlight;
    }
    if (maxInFlightBytesSizeEnabled_ && batchBytesSize) {
      if (totalInflightBytes_.load() > maxInFlightBytesSizeTotal_ ||
          batchBytesSize->load() > maxInFlightBytesSizePerWorker_) {
        LOG(WARNING) << "After waiting for " << waitInflightTimeoutMs_
                     << " ms, there are still " << totalInflightBytes_.load()
                     << " bytes in flight (limit: "
                     << maxInFlightBytesSizeTotal_
                     << "): " << batchBytesSize->load()
                     << " bytes for hostAndPushPort " << hostAndPushPort
                     << ", which exceeds the current limit "
                     << maxInFlightBytesSizePerWorker_;
      }
    }
  }
  throwIfExceptionExists();
  return times <= 0;
}

bool PushState::limitZeroInFlight() {
  throwIfExceptionExists();

  long times = waitInflightTimeoutMs_ / deltaMs_;
  for (; times > 0; times--) {
    if (cleaned_) {
      return false;
    }
    if (totalInflightReqs_ <= 0) {
      break;
    }
    throwIfExceptionExists();
    std::this_thread::sleep_for(utils::MS(deltaMs_));
  }

  if (times <= 0) {
    std::string addressInfos;
    inflightBatchesPerAddress_.forEach(
        [&](const std::string& address,
            const std::shared_ptr<utils::ConcurrentHashSet<int>>&
                inflightBatches) {
          if (inflightBatches->size() <= 0) {
            return;
          }
          if (!addressInfos.empty()) {
            addressInfos += ", ";
          }
          addressInfos += fmt::format(
              "{} batches for hostAndPushPort {}",
              inflightBatches->size(),
              address);
        });
    LOG(ERROR) << "After waiting for " << waitInflightTimeoutMs_
               << " ms, there are still " << totalInflightReqs_
               << " in flight: [" << addressInfos
               << "] which exceeds the current limit 0.";
  }
  throwIfExceptionExists();
  return times <= 0;
}

bool PushState::exceptionExists() const {
  auto exp = exception_.rlock();
  return (bool)(*exp);
}

void PushState::setException(std::unique_ptr<std::exception> exception) {
  auto exp = exception_.wlock();
  if (!(*exp)) {
    *exp = std::move(exception);
  }
}

std::optional<std::string> PushState::getExceptionMsg() const {
  auto exp = exception_.rlock();
  if (*exp) {
    return (*exp)->what();
  }
  return std::nullopt;
}

void PushState::cleanup() {
  LOG(INFO) << "Cleanup " << totalInflightReqs_.load()
            << " requests in flight.";
  cleaned_ = true;
  inflightBatchesPerAddress_.clear();
  totalInflightReqs_ = 0;
  pushStrategy_->clear();

  if (maxInFlightBytesSizeEnabled_) {
    LOG(INFO) << "Cleanup " << totalInflightBytes_.load()
              << " bytes in flight.";
    inflightBytesSizePerAddress_->clear();
    inflightBatchBytesSizes_->clear();
    totalInflightBytes_ = 0;
  }
}

void PushState::throwIfExceptionExists() {
  auto exp = exception_.rlock();
  if (*exp) {
    CELEBORN_FAIL((*exp)->what());
  }
}

} // namespace client
} // namespace celeborn
